test/unit/math/prim/fun/as_value_array_or_scalar_test.cpp

```diff
@@ -0,0 +1,63 @@
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/as_value_array_or_scalar.hpp>
+#include <test/unit/util.hpp>
+#include <gtest/gtest.h>
+
+TEST(MathMetaPrim, as_value_array_or_scalar_scalar) {
+  int a = 3;
+  double b = 4;
+  EXPECT_EQ(a, stan::math::as_value_array_or_scalar(a));
+  EXPECT_EQ(b, stan::math::as_value_array_or_scalar(b));
+}
+
+TEST(MathMetaPrim, as_value_array_or_scalar_std_vector_lvalue) {
+  int n = 100;
+  Eigen::ArrayXd a = Eigen::ArrayXd::Random(n);
+  std::vector<double> b(n);
+  for (int i = 0; i < n; i++) {
+    b[i] = a.coeff(i);
+  }
+  const auto& tmp = stan::math::as_value_array_or_scalar(b);
+  Eigen::ArrayXd res = tmp;
```

Comment:
Add in some simple type checks to all the tests too, since that's a big part of what is happening.

Like:

```EXPECT_TRUE((stan::math::is_eigen_array<decltype(tmp)>::value))```